From 37fa9c01d27328c3ab23bcc7e554a2ceac937990 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Thu, 27 Nov 2014 11:46:53 -0500
Subject: [fs] ovl: allow filenames with comma

Message-id: <20141127114653.9170.44087.stgit@warthog.procyon.org.uk>
Patchwork-id: 100987
O-Subject: [RHEL7 PATCH 4/7] ovl: allow filenames with comma
Bugzilla: 985875
RH-Acked-by: Alexander Viro <aviro@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

Allow option separator (comma) to be escaped with backslash.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Original-commit: 91c77947133f7aef851b625701e182d3f99d14a9
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/overlayfs/super.c b/fs/overlayfs/super.c
index aad1e9e..74fa03b 100644
--- a/fs/overlayfs/super.c
+++ b/fs/overlayfs/super.c
@@ -462,11 +462,34 @@ static const match_table_t ovl_tokens = {
  {OPT_ERR,   NULL}
 };
 
+static char *ovl_next_opt(char **s)
+{
+ char *sbegin = *s;
+ char *p;
+
+ if (sbegin == NULL)
+  return NULL;
+
+ for (p = sbegin; *p; p++) {
+  if (*p == '\\') {
+   p++;
+   if (!*p)
+    break;
+  } else if (*p == ',') {
+   *p = '\0';
+   *s = p + 1;
+   return sbegin;
+  }
+ }
+ *s = NULL;
+ return sbegin;
+}
+
 static int ovl_parse_opt(char *opt, struct ovl_config *config)
 {
  char *p;
 
- while ((p = strsep(&opt, ",")) != NULL) {
+ while ((p = ovl_next_opt(&opt)) != NULL) {
   int token;
   substring_t args[MAX_OPT_ARGS];
 
@@ -554,15 +577,34 @@ out_dput:
  goto out_unlock;
 }
 
+static void ovl_unescape(char *s)
+{
+ char *d = s;
+
+ for (;; s++, d++) {
+  if (*s == '\\')
+   s++;
+  *d = *s;
+  if (!*s)
+   break;
+ }
+}
+
 static int ovl_mount_dir(const char *name, struct path *path)
 {
  int err;
+ char *tmp = kstrdup(name, GFP_KERNEL);
+
+ if (!tmp)
+  return -ENOMEM;
 
- err = kern_path(name, LOOKUP_FOLLOW, path);
+ ovl_unescape(tmp);
+ err = kern_path(tmp, LOOKUP_FOLLOW, path);
  if (err) {
-  pr_err("overlayfs: failed to resolve '%s': %i\n", name, err);
+  pr_err("overlayfs: failed to resolve '%s': %i\n", tmp, err);
   err = -EINVAL;
  }
+ kfree(tmp);
  return err;
 }
 
-- 
1.7.1